### PR TITLE
CLI: Raise if no root node specified in `verdi node graph generate`

### LINKED ISSUE #6397: `verdi node graph generate` error for missing PK is not correctly handled and help page is wrong
### Describe the bug / Steps to reproduce / Expected behavior

- Error for missing is PK not properly handled
```bash
$ verdi node graph generate
Traceback (most recent call last):
  File "/home/alexgo/micromamba/envs/aiida-dev/bin/verdi", line 8, in <module>
    sys.exit(verdi())
             ^^^^^^^
  File "/home/alexgo/micromamba/envs/aiida-dev/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alexgo/micromamba/envs/aiida-dev/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/alexgo/micromamba/envs/aiida-dev/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alexgo/micromamba/envs/aiida-dev/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alexgo/micromamba/envs/aiida-dev/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alexgo/micromamba/envs/aiida-dev/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alexgo/micromamba/envs/aiida-dev/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alexgo/code/aiida-core/src/aiida/cmdline/utils/decorators.py", line 102, in wrapper
    return wrapped(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alexgo/code/aiida-core/src/aiida/cmdline/commands/cmd_node.py", line 535, in graph_generate
    pks = '.'.join(str(n.pk) for n in root_nodes)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alexgo/code/aiida-core/src/aiida/cmdline/commands/cmd_node.py", line 535, in <genexpr>
    pks = '.'.join(str(n.pk) for n in root_nodes)
                       ^^^^
AttributeError: 'NoneType' object has no attribute 'pk'
```
Compare to
```bash
$ verdi computer show
Usage: verdi computer show [OPTIONS] COMPUTER
Try 'verdi computer show --help' for help.

Error: Missing argument 'COMPUTER'.
```
The error seems not to be properly handled in the former case.

- The help page of the command is wrong
```bash
$ verdi node graph generate --help
Usage: verdi node graph generate [OPTIONS] [--] [ROOT_NODE] OUTPUT_FILE

  Generate a graph from one or multiple root nodes.
[...]
```
Again comparing it with other commands, I interpret `[ROOT_NODE]` as optional, but it is not, it is the `OUTPUT_FILE` which is optional. Also `verdi node graph generate -N 1 output.pdf` does not work but one has to do `verdi node graph generate -N 1 -O output.pdf`

### Your environment

- Operating system: Ubuntu 22.04 LTS
- Python version: 3.11
```
$ verdi status
✔ version:     AiiDA v2.5.1
✔ config:      /home/alexgo/code/aiida-core/.aiida
✔ profile:     alexgo
✔ storage:     SqliteDosStorage[/home/alexgo/code/aiida-core/.aiida/repository/sqlite_dos_f275ff0f10174e8e84c13e82dd5ba452]: open,
✔ rabbitmq:    Connected to RabbitMQ v3.8.14 as amqp://guest:guest@127.0.0.1:5672?heartbeat=600
```

### PR DESCRIPTION
**Pull request by sphuber**
Fixes #6397

The `root_node` positional argument was recently made optional since it was deprecated and replaced by the `-N/--nodes` option. However, the change forgot to check that at least one root node is specified. Not specifying any would result in an exception.

### CODE REVIEW

**Comment by sphuber** [comment]
Labeling as blocking since this bug was introduced by changes that have not yet been released

File: src/aiida/cmdline/commands/cmd_node.py [hunk 1]
```diff
@@ -510,6 +510,9 @@ def graph_generate(
         )
         root_node = None
 
+    if not root_node and not nodes:
+        echo.echo_critical('No root node specified, please specify one or more using the `-N/--nodes` option.')
+
     if root_node:
         echo.echo_deprecated(
```
Comment (danielhollas) [inline src/aiida/cmdline/commands/cmd_node.py:517]: Unrelated to the fix, but is it really necessary to deprecate the original behaviour? Supposedly this is the most common use case, so why not allow this shorthand.

(Apologies if this was already discussed ignore me in that case)

Comment (sphuber) [reply src/aiida/cmdline/commands/cmd_node.py:517]: Because it would be potentially confusing if there are two methods. If we were to keep it, the user wanting to specify multiple nodes would still have to specify the root node positional argument, e.g., `verdi node graph generate -N 10 11 -- 15 output.pdf`. I guess we could either change the order of `ROOT_NODE` and `OUTPUT_FILE` or change the output file from an argument to an option. But in both cases we would be changing the interface.

One other solution perhaps (if we are thinking of changing the interface anyway) would be to get rid of the `-N` option, turn output file from argument to an option, and allow root nodes to be specified as positional arguments. So example uses would be
```
verdi node graph generate 10
verdi node graph generate 10 11
verdi node graph generate -o output.pdf 10
verdi node graph generate -o output.pdf 10 11
```
Maybe that would be cleanest?

Most of the `verdi node` commands that take more than one node use arguments. But there are also examples where the `-N/--nodes` option is used, but those are limited. I think in most cases using arguments would make most sense. The reason `verdi archive create` uses an option is because you can also specify groups and other entities, so there it makes sense. Maybe using arguments in `verdi node graph generate` really makes most sense.

This is what `click` has to say [in their docs](https://click.palletsprojects.com/en/8.1.x/parameters/) about options vs arguments:

> Click supports two types of parameters for scripts: options and arguments. There is generally some confusion among authors of command line scripts of when to use which, so here is a quick overview of the differences. As its name indicates, an option is optional. While arguments can be optional within reason, they are much more restricted in how optional they can be.

> To help you decide between options and arguments, the recommendation is to use arguments exclusively for things like going to subcommands or input filenames / URLs, and have everything else be an option instead.

The first paragraph would suggest we use an argument since at least one root node is required and not optional. However, the second paragraph would suggest using an option, but maybe our use case is a bit more specific compared to most CLI tools. The output filename would seem to fit in the argument category, but could also be argued to be added as an option, since the command will generate an output file by default.

TLDR: Maybe the best solution would be to move output file to an option and specify the root nodes with arguments?

Comment (sphuber) [reply src/aiida/cmdline/commands/cmd_node.py:517]: @danielhollas I opened a new category on discourse for design discussions. Opened a first topic on the question in this PR. Seems that people are gravitating to using arguments to allow specifying root nodes and use an option for the optional output file. Let us know what you think please: https://aiida.discourse.group/t/redesign-of-the-verdi-node-graph-generate-command/395

Comment (danielhollas) [reply src/aiida/cmdline/commands/cmd_node.py:517]: Replied on Discourse, thanks!

**Review by agoscinski — APPROVED** [review]
Thanks for the fix! Just a rebase needed

**Comment by sphuber** [comment]
Thanks for the review @agoscinski but given the responses on the [discourse thread](https://aiida.discourse.group/t/redesign-of-the-verdi-node-graph-generate-command/395/4) it seems like we would rather be going with a different solution, as implemented in https://github.com/aiidateam/aiida-core/pull/6427
If you have the time to review that, that would be great.

**Comment by sphuber** [comment]
Superseded by #6427
